### PR TITLE
[sourcekit] Fix for Ubuntu 16.04 Swift ASAN bot sourcekitd-test failures

### PR DESCRIPTION
**Pull request by nathawes**
https://ci.swift.org/job/oss-swift-incremental-ASAN-RA-ubuntu-16_04/

All the `sourcekitd-test` invocations are failing due to:
`sourcekitd-test: symbol lookup error: /home/buildnode/jenkins/workspace/oss-swift-incremental-ASAN-RA-ubuntu-16_04/buildbot_incremental_asan/swift-linux-x86_64/tools/SourceKit/libdispatch-prefix/lib/libBlocksRuntime.so: undefined symbol: __asan_version_mismatch_check_v8`

The change here forces using the host compiler if sanitizers are enabled.

Resolves rdar://problem/45273748

### CODE REVIEW

**Comment by nathawes** [comment]
@swift-ci please test

**Comment by nathawes** [comment]
@swift-ci please test

File: tools/SourceKit/CMakeLists.txt [hunk 1]
```diff
@@ -100,8 +100,11 @@ if(NOT "${CMAKE_SYSTEM_NAME}" STREQUAL "Darwin")
   set(SOURCEKIT_NEED_EXPLICIT_LIBDISPATCH TRUE)
 
   if(SWIFT_BUILD_SOURCEKIT)
+    # NOTE always use the just-built clang for ASAN builds to avoid
+    # compatibility issues when components are built with different compilers
```
[the remark this answers is not in the document]
Comment (nathawes) [reply tools/SourceKit/CMakeLists.txt]: Sorry, I probably shouldn't have commented at all. I don't really understand what's happening at this point beyond the ASAN version mismatch. That usually means two different compilers were used on the asanified pieces, and the commit message on https://github.com/apple/swift/commit/ed3cb02475efc7a7d9c63c43add31d8004624d2e made it seem like that change could've triggered the failures.  The condition change below is just intended as a revert of that commit that only applies to the ASAN bots (the elseif branch matches the previous behavior) to see if it's the issue or not.

If you have time to look into the bot failure or have any ideas what might be causing it that would be appreciated! I'm setting up a repro environment at the moment to dig into it more.

**Comment by nathawes** [comment]
@compnerd I got back to this and managed to reproduce the failure locally with an Ubuntu 16.04 docker image and you were right, the check was the wrong way around. The changes as they stand now fixed the ASAN mismatch in `sourcekitd-test` locally for me, but I was wondering if we could lower the required version here generally instead of forcing it to use the host compiler when ASAN is enabled.

Was there a specific requirement for 3.9+ building libdispatch here or was this just a known good configuration? It seems to build fine when I drop it to > 3.7 locally (my Ubuntu 16.04 docker setup has 3.8).

**Comment by nathawes** [comment]
@swift-ci please test

**Comment by nathawes** [comment]
@swift-ci please smoke test osx

**Comment by nathawes** [comment]
@swift-ci smoke test osx

**Comment by nathawes** [comment]
@swift-ci please smoke test

**Comment by shahmishal** [comment]
@nathawes We are going kill Linux build to investigate lldb test issue.

**Comment by nathawes** [comment]
@swift-ci Please smoke test Linux platform

**Pull request merged**